docs: Azure Disk Encryption (PCP-3720 & PCP-3735) (#4558)
spectrocloud/librarium: docs/docs-content/clusters/public-cloud/azure/azure-disk-encryption.md

Reviewer: @benradstone I think there is some confusion about the required sections for how-tos. Each how-to, requires its own prereq, enablement (steps) section, and validation. In this case, you have have multiple how-tos in a single page, which is fine as it's all related to the same topic. But we still need to honor the layout guide.  Here is an [example](https://docs.spectrocloud.com/clusters/cluster-management/backup-restore/add-backup-location-static/) of a guide with multiple how-tos.

TLDR: Treat each how-to as it's own section. Trying to combine a validation and prereq for multiple topics gets confusing from a reader perspective. 

Another minor thing I caught a few times was incorrect UI element labeling. Take a peek at the Product UI Naming table of the[ style guide](https://spectrocloud.atlassian.net/wiki/spaces/DE/pages/1765933057/Spectro+Cloud+Internal+Style+Guide).

---
sidebar_label: "Azure Disk Encryption"
title: "Azure Disk Encryption"
description:
  "How to enable disk encryption on your Azure IaaS cluster using Disk Encryption Sets with customer-managed keys"
icon: ""
hide_table_of_contents: false
tags: ["public cloud", "azure", "encryption", "security"]
sidebar_position: 12
---

> Palette supports disk encryption of your Azure Kubernetes cluster using Disk Encryption Sets with

Author: I'll adjust the hyperlink to start from here, rather than just Customer-managed keys.

> Palette supports disk encryption of your Azure Kubernetes cluster using Disk Encryption Sets with
> [customer-managed keys](https://learn.microsoft.com/en-us/azure/virtual-machines/disk-encryption#customer-managed-keys).

Author: ```suggestion
Palette supports disk encryption of your Azure Kubernetes cluster using [Disk Encryption Sets with
customer-managed keys](https://learn.microsoft.com/en-us/azure/virtual-machines/disk-encryption#customer-managed-keys).
```

By default, Azure encrypts all managed disks with
[platform-managed keys](https://learn.microsoft.com/en-us/azure/virtual-machines/disk-encryption#platform-managed-keys),
however, customer-managed keys enable you to have greater control over your key management.

You can use Disk Encryption Sets to encrypt your nodes' Operating System (OS) and data disks by including a preset
option in the <VersionedLink text="Palette eXtended Kubernetes (PXK)" url="/integrations/packs/?pack=kubernetes" />
pack.

## Limitations

> - Azure Disk Encryption is available from [4.5.a](../../../release-notes/release-notes.md) onwards.

Reviewer: You don't need to do this as our docs are versioned. This content will only be available in new versions of our docs. 

```suggestion
```

- Azure Disk Encryption is only supported on Azure IaaS clusters.

- Azure Disk Encryption is only supported when using the

>   <VersionedLink text="Palette eXtended Kubernetes (PXK)" url="/integrations/packs/?pack=kubernetes" /> pack.

Reviewer: Fix prettier newline

> - Azure Disk Encryption is only supported when using the
> 
>   <VersionedLink text="Palette eXtended Kubernetes (PXK)" url="/integrations/packs/?pack=kubernetes" /> pack.

Author: ```suggestion
  <!-- prettier-ignore -->
- Azure Disk Encryption is only supported when using the
  <VersionedLink text="Palette eXtended Kubernetes (PXK)" url="/integrations/packs/?pack=kubernetes" /> pack.
```

- If a key expires in your Key Vault, your cluster may experience operation failures. To resolve this, generate a new
  key in Key Vault and update your Disk Encryption Set to reference the new key.

  - No changes are needed in Palette, as the Palette eXtended Kubernetes pack configuration references the URI of your
    Disk Encryption Set, which remains unchanged.

  - We recommend enabling
    [**Auto key rotation**](https://learn.microsoft.com/en-us/azure/virtual-machines/disk-encryption#automatic-key-rotation-of-customer-managed-keys)
    on your Disk Encryption Set so it can automatically use new key versions from your Key Vault.

- Changing the Disk Encryption Set URI (`diskEncryptionSetID`) in the Palette eXtended Kubernetes pack configuration in
  Palette will trigger a node repave.

## Prerequisites

- To create the Azure Key Vault and Disk Encryption Set, you will need an Azure user account with the **Contributor** or

>   **Owner** role on the resource group or subscription.

Reviewer: We do not recommend the contributor role. We have gone to great lengths to document all the required Azure [permissions](https://deploy-preview-4558--docs-spectrocloud.netlify.app/clusters/public-cloud/azure/required-permissions/) so using the contributor is frowned upon. So if there are permission they need then let's call them out, but let's not tell people to use the overprivileged Contributor role because they will use it a point back to this doc.

Author: They mention this in the Feature Overview, but I will break this down into permissions required for each task.

- Create an
  [Azure Key Vault](https://learn.microsoft.com/en-us/azure/virtual-machines/disks-enable-customer-managed-keys-portal#set-up-your-azure-key-vault)
  with the following necessary configuration:

  - Resource access:

>     - :white_check_mark: **Azure Virtual Machines for deployment**

Author: No, I should have read our Emoticons section better in our style guide.

  - Purge protection:
    - :white_check_mark: **Enable purge protection**

- Create a
  [Disk Encryption Set](https://learn.microsoft.com/en-us/azure/virtual-machines/disks-enable-customer-managed-keys-portal#set-up-your-disk-encryption-set)
  with the encryption type as **Encryption at-rest with a customer-managed key**.

- In the Azure Key Vault, assign the following
  [access policies](https://learn.microsoft.com/en-us/azure/key-vault/general/assign-access-policy?tabs=azure-portal) to
  the Disk Encryption Set that you want to use:

  - Key Management Operations:

    - :white_check_mark: **Get**

  - Cryptographic Operations:

    - :white_check_mark: **Unwrap Key**
    - :white_check_mark: **Wrap Key**

> 

Reviewer: @benradstone I'm not sure what this is, but it's not our style. The emojis are primarily only reserved for markdown tables with a small set of exceptions.  Consider using a table to list out the required permissions for each resource, or stick with a list. Here is an [example](https://docs.spectrocloud.com/clusters/public-cloud/aws/enable-secrets-encryption-kms-key/) where IAM permissions are specified for a how to.

Author: Thank you, I was looking for an example like that.

  - If you have designated a user-assigned identity to the Disk Encryption Set, add the same access policies to the
    user-assigned identity in the Azure Key Vault.

## Enable Disk Encryption

### New Cluster Profile

Use the following steps to enable disk encryption on a
[new cluster profile](../../../profiles/cluster-profiles/create-cluster-profiles/create-cluster-profiles.md).

1. Log in to [Palette](https://console.spectrocloud.com).

2. Ensure you are in the correct project scope.

3. From the left **Main Menu**, select **Profiles** and click **Add Cluster Profile**.

4. Fill out the **Basic Information** and ensure **Type** is not set to **Add-on**. Click **Next**.

5. In **Cloud Type**, under **Infrastructure provider**, select **Azure IaaS**. Click **Next**.

6. Select your base OS pack and configure it to your liking. Select **Next layer** when complete.

7. Select the **Palette eXtended Kubernetes** pack as your base Kubernetes pack.

8. In **Configure Pack**, select **Values** in pack details, and click the **\</\>** button to show the YAML editor.

> 9. On the right-hand side, click the **Presets** drop-down menu, and select the **Enable Encryption Using

Reviewer: ```suggestion
9. On the right-hand side, click the **Presets** drop-down Menu, and select the **Enable Encryption Using
```

   Customer-Managed Key** option.

10. Scroll to the bottom of the YAML editor to view the additional configuration that was added.

    ```yaml
    cloud:
      azure:
        diskEncryptionSetID: ""
    ```

11. Fill in the `diskEncryptionSetID` with the Resource ID URI of your Disk Encryption Set.

    <!-- prettier-ignore -->
    <details>
    <summary> How to find the Resource ID URI of your Disk Encryption Set </summary>

    1. Log in to the [Azure Portal](https://portal.azure.com/).

>     2. In the search bar, look for **Disk Encryption Sets**. Click on the service when found.

Reviewer: Let's use more explicit language so that the user knows to click on the search bar and type out the Disk query.

    3. Find your Disk Encryption Set from the list and click on it to view details.

    4. On the **Overview** page, click **JSON View** in the **Essentials** section. The Resource ID for the Disk
       Encryption Set is displayed at the top.

>     5. Click the **Copy to clipboard** icon for the Resource ID and paste it into the `diskEncryptionSetID` field in the

Author: I wrote it as an implication that the user will see their subscriptionId and resourceGroup in the full URI.

The Resource ID contains the full URI (including the Sub and RG IDs), so to avoid confusion, I'll remove the edged-brackets.

       Palette YAML editor.

    </details>

    ```yaml
    cloud:
      azure:

>         diskEncryptionSetID: "/subscriptions/<subscriptionId>/resourceGroups/<resourceGroup>/providers/Microsoft.Compute/diskEncryptionSets/<diskEncryptionSet>"

Author: ```suggestion
        diskEncryptionSetID: "/subscriptions/subscriptionId/resourceGroups/resourceGroup/providers/Microsoft.Compute/diskEncryptionSets/diskEncryptionSet"
```

    ```

12. Make any other changes that you need and click **Next layer**.

13. Select the remaining profile layers to finish the configuration.

You can now [create a new Azure IaaS cluster](./create-azure-cluster.md) with disk encryption enabled using this cluster
profile.

> ### Edit an Existing Cluster Profile

Reviewer: Every how-to needs a prerequisites section and a validation section.

#### Enable Disk Encryption on a Cluster Profile

Use the following steps to enable disk encryption on an
[existing cluster profile](../../../profiles/cluster-profiles/modify-cluster-profiles/modify-cluster-profiles.md).

1. Log in to [Palette](https://console.spectrocloud.com).

2. Ensure you are in the correct project scope.

3. From the left **Main Menu**, select **Profiles** and click the cluster profile that you want to edit.

4. Create a new version of your cluster profile:

   1. Click the version drop-down menu next to the cluster profile name, and click **Create new version**.
   2. Fill the **Version** field with a new version number.
   3. Click **Confirm**.

5. Select the Kubernetes layer to view the **Edit Pack** drawer.

   :::warning

   Ensure that you are using

>    <VersionedLink text="Palette eXtended Kubernetes (PXK)" url="/integrations/packs/?pack=kubernetes" /> as the Azure

Reviewer: Same feedback as earlier, use the prettier ignore to fix the newline injected by prettier.
But, also, this can be addressed by making it a prereq, so we could do without an admonition.

Author: We do call it out in limitations. Not sure if this suffices or not.

>    Ensure that you are using
> 
>    <VersionedLink text="Palette eXtended Kubernetes (PXK)" url="/integrations/packs/?pack=kubernetes" /> as the Azure
>    Disk Encryption option is only available to this pack.

Author: ```suggestion
   <!-- prettier-ignore -->
   Ensure that you are using
   <VersionedLink text="Palette eXtended Kubernetes (PXK)" url="/integrations/packs/?pack=kubernetes" />, as the Azure
   Disk Encryption option is only available to this pack.
```

   :::

6. In **Values**, click the **\</\>** button to show the YAML editor.

7. On the right-hand side, click the **Presets** drop-down menu, and select the **Enable Encryption Using
   Customer-Managed Key** option.

8. Scroll to the bottom of the YAML editor to view the additional configuration that was added.

   ```yaml
   cloud:
     azure:
       diskEncryptionSetID: ""
   ```

9. Fill in the `diskEncryptionSetID` with the Resource ID URI of your Disk Encryption Set.

   <!-- prettier-ignore -->
   <details>
   <summary> How to find the Resource ID URI of your Disk Encryption Set </summary>

   1. Log in to the [Azure Portal](https://portal.azure.com/).

   2. In the search bar, look for **Disk Encryption Sets**. Click on the service when found.

   3. Find your Disk Encryption Set from the list and click on it to view details.

   4. On the **Overview** page, click **JSON View** in the **Essentials** section. The Resource ID for the Disk
      Encryption Set is displayed at the top.

>    5. Click the **Copy to clipboard** icon for the Resource ID and paste it into the `diskEncryptionSetID` field in the

Reviewer 2: we should also tell users to replace `<subscriptionId>` and `<resourceGroup>` with their information

      Palette YAML editor.

   </details>

   ```yaml
   cloud:
     azure:

>        diskEncryptionSetID: "/subscriptions/<subscriptionId>/resourceGroups/<resourceGroup>/providers/Microsoft.Compute/diskEncryptionSets/<diskEncryptionSet>"

Author: ```suggestion
       diskEncryptionSetID: "/subscriptions/subscriptionId/resourceGroups/resourceGroup/providers/Microsoft.Compute/diskEncryptionSets/diskEncryptionSet"
```

   ```

10. Click **Confirm updates**.

11. Click **Save Changes**.

> #### Enable Disk Encryption on an Active Cluster

Reviewer 2: I think we could merge the  `Enable Disk Encryption on a Cluster Profile` and `Enable Disk Encryption on an Active Cluster` sub guides, once one is a continuation of the other. Then, we could have only one h3 heading `Enable Disk Encryption on an Active Cluster` . What do you think?

Author: I thought about doing this originally. I kept them separate to call out the admonition warning before a user started this part, but it makes sense to merge them together.

I'll call it out either at the beginning, or before they move onto these steps instead.

Use the following steps to enable disk encryption on an active cluster using the
[updated cluster profile](#enable-disk-encryption-on-a-cluster-profile).

> :::warning

Reviewer: Good use of warning

[full cluster repave](../../cluster-management/node-pool.md#repave-behavior-and-configuration).

:::

1. From the left **Main Menu**, select **Clusters**.

2. Find the cluster that you want to update and click on it.

3. Click the **Profile** tab.

4. Click the version drop-down menu in **Infrastructure Layers** and select the version that has disk encryption
   enabled.

5. Click **Review & Save**, then click **Review changes in Editor** in the Changes Summary box.

6. Review the changes and click **Update**.

Your cluster will now update and a full cluster repave will occur. Wait until the update has completed before
[validating the disk encryption enablement](#validate-after-enablement).

## Disable Disk Encryption

### Disable Disk Encryption on a Cluster Profile

Use the following steps to disable disk encryption on an
[existing cluster profile](../../../profiles/cluster-profiles/modify-cluster-profiles/modify-cluster-profiles.md).

1. Log in to [Palette](https://console.spectrocloud.com).

2. Ensure you are in the correct project scope.

3. From the left **Main Menu**, select **Profiles** and click the cluster profile that you want to edit.

> 4. Create a new version of your cluster profile:

Reviewer 2: Same comment about sup-steps

>    1. Click the version drop-down menu next to the cluster profile name, and click **Create new version**.

Reviewer: ```suggestion
   1. Click the version **drop-down Menu** next to the cluster profile name, and click **Create new version**.
```

Drop-down Menu should be in bold as its a UI element.

   Customer-Managed Key** option.

8. Click **Confirm updates**.

9. Click **Save Changes**.

### Disable Disk Encryption on an Active Cluster

Use the following steps to disable disk encryption on an active cluster using the
[updated cluster profile](#disable-disk-encryption-on-a-cluster-profile).

:::warning

Performing these steps will cause a
[full cluster repave](../../cluster-management/node-pool.md#repave-behavior-and-configuration).

:::

1. From the left **Main Menu**, select **Clusters**.

2. Find the cluster that you want to update and click on it.

3. Click the **Profile** tab.

4. Click the version drop-down menu in **Infrastructure Layers** and select the version that has disk encryption
   disabled.

5. Click **Review & Save**, then click **Review changes in Editor** in the Changes Summary box.

6. Review the changes and click **Update**.

Your cluster will now update and a full cluster repave will occur. Wait until the update has completed before
[validating the disk encryption disablement](#validate-after-disablement).

## Validate

### Validate after Enablement

> Follow these steps to validate enablement of customer-managed key encryption on your Azure VM disks.

Reviewer 2: ```suggestion
Follow these steps to validate the enablement of customer-managed key encryption on your Azure VM disks.
```

1. Log in to the [Azure Portal](https://portal.azure.com/).

2. In the search bar, look for **Disks**. Click on the service when found.

3. Find your cluster's disks by using the search filters provided. You can use the cluster name as the cluster resources
   contain the cluster name at the beginning (for example: `<clusterName>-e3c0-f7ljd_OSDisk`).

4. Once identified, click on a disk name to view its details.

> 5. Scroll down to view the **Properties** tab and check that the **Encryption** section shows **Customer-managed key**

Reviewer 2: Maybe we could merge the `Validate after Enablement` and `Validate after Disablement` guides as well to simplify? Step 5 is the only different step

   for **Encryption type**. Your Disk Encryption Set name is also shown for the **Encryption key**.

> 6. Repeat steps 4-5 for each disk in your cluster.

Reviewer 2: ```suggestion
6. Repeat steps four and five for each disk in your cluster.
```

Author: ```suggestion
6. Repeat steps 4 and 5 for each disk in your cluster.
```

Author: @caroldelwing In our style-guide, we stipulate:

> In how-to steps, if instructions refer to a specific step or steps, use numbers instead of spelling out the word.

### Validate after Disablement

> Follow these steps to validate disablement of customer-managed key encryption on your Azure VM disks.

Reviewer 2: ```suggestion
Follow these steps to validate the disablement of customer-managed key encryption on your Azure VM disks.
```

1. Log in to the [Azure Portal](https://portal.azure.com/).

2. In the search bar, look for **Disks**. Click on the service when found.

3. Find your cluster's disks by using the search filters provided. You can use the cluster name as the cluster resources
   contain the cluster name at the beginning (for example: `<clusterName>-e3c0-f7ljd_OSDisk`).

4. Once identified, click on a disk name to view its details.

5. Scroll down to view the **Properties** tab and check that the **Encryption** section shows **Platform-managed key**
   for **Encryption type**.

> 6. Repeat steps 4-5 for each disk in your cluster.

Reviewer 2: ```suggestion
6. Repeat steps four and five for each disk in your cluster.
```

- Azure Disk Encryption is only supported when using the <VersionedLink text="Palette eXtended Kubernetes (PXK)" url="/integrations/packs/?pack=kubernetes" /> pack.

- If a key expires in your Key Vault, your cluster may experience operation failures. To resolve this, generate a new
  key in Key Vault and update your Disk Encryption Set to reference the new key.

  - No changes are needed in Palette, as the Palette eXtended Kubernetes pack configuration references the URI of your
    Disk Encryption Set, which remains unchanged.

  <!-- prettier-ignore -->
  - We recommend enabling
    [**Auto key rotation**](https://learn.microsoft.com/en-us/azure/virtual-machines/disk-encryption#automatic-key-rotation-of-customer-managed-keys)
    on your Disk Encryption Set so it can automatically use new key versions from your Key Vault.

- Changing the Disk Encryption Set URI (`diskEncryptionSetID`) in the Palette eXtended Kubernetes pack configuration in
  Palette will trigger a node repave.

## New Cluster Profile

### Prerequisites {#prerequisites-new-cluster-profile}

- An Azure user account with the following roles to create the Azure Key Vault and Disk Encryption Set with the

>   necessary Key Vault access policies.

Reviewer: Might be good to link to the Azure docs explanation of the built in key roles
https://learn.microsoft.com/en-us/azure/key-vault/general/rbac-guide?tabs=azure-cli#azure-built-in-roles-for-key-vault-data-plane-operations

- Changing the Disk Encryption Set URI in the Palette eXtended Kubernetes pack configuration in
  Palette will trigger a node repave.

## New Cluster Profile

### Prerequisites {#prerequisites-new-cluster-profile}

- An Azure user account with the following roles to create the Azure Key Vault and Disk Encryption Set with the
  necessary Key Vault access policies.

  | Task                                                                          | Required Role                            |
  | ----------------------------------------------------------------------------- | ---------------------------------------- |
  | Create Key Vault and Key                                                      | Key Vault Contributor                    |
  | Create Disk Encryption Set & Assign Key Vault Key                             | Key Vault Administrator                  |
  | Assign Key Vault Access Policies (GET, WRAP KEY, UNWRAP KEY)                  | Key Vault Crypto Service Encryption User |
  | (Optional) Assign User-Assigned Identity to Key Vault and Disk Encryption Set | Managed Identity Operator                |

>   For more information, see

Author: ```suggestion
  For more information, visit
```

---
sidebar_label: "Azure Disk Encryption"
title: "Azure Disk Encryption"
description:
  "How to enable disk encryption on your Azure IaaS cluster using Disk Encryption Sets with customer-managed keys"
icon: ""
hide_table_of_contents: false
tags: ["public cloud", "azure", "encryption", "security"]
sidebar_position: 12
---

Palette supports disk encryption of your Azure Kubernetes cluster using
[Disk Encryption Sets with customer-managed keys](https://learn.microsoft.com/en-us/azure/virtual-machines/disk-encryption#customer-managed-keys).

By default, Azure encrypts all managed disks with
[platform-managed keys](https://learn.microsoft.com/en-us/azure/virtual-machines/disk-encryption#platform-managed-keys),
however, customer-managed keys enable you to have greater control over your key management.

> <!-- prettier-ignore -->
> You can use Disk Encryption Sets to encrypt your nodes' Operating System and data disks by selecting a preset in the
> 
> <VersionedLink text="Palette eXtended Kubernetes" url="/integrations/packs/?pack=kubernetes" /> pack.

Author: ```suggestion
<!-- prettier-ignore -->
You can use Disk Encryption Sets to encrypt your nodes' Operating System and data disks by selecting a preset in the <VersionedLink text="Palette eXtended Kubernetes" url="/integrations/packs/?pack=kubernetes" /> pack.
```
